subprojects/dependency-management/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java

```diff
@@ -157,55 +163,74 @@ public boolean isMatching(AttributeContainerInternal candidate, AttributeContain
 
         @Override
         public boolean hasAttribute(Attribute<?> attribute) {
-            return getAttributes().contains(attribute) || producerSchema.getAttributes().contains(attribute);
+            return DefaultAttributesSchema.this.getAttributes().contains(attribute) || producerSchema.getAttributes().contains(attribute);
         }
 
         @Override
-        public void disambiguate(Attribute<?> attribute, MultipleCandidatesResult<Object> result) {
+        public Set<Object> disambiguate(Attribute<?> attribute, Object requested, Set<Object> candidates) {
+            DefaultMultipleCandidateResult<Object> result = null;
+
             DisambiguationRule<Object> rules = disambiguationRules(attribute);
-            rules.execute(result);
-            if (result.hasResult()) {
-                return;
+            if (rules.doesSomething()) {
+                result = new DefaultMultipleCandidateResult<Object>(requested, candidates);
```

Comment:
This exact same block of code appears 4 times, maybe extract?

Reply:
Unfortunately not without reintroducing the garbage creation that this change avoids.